### PR TITLE
Added plugin Directory for Benchmarks

### PR DESCRIPTION
**Pull request by JBenda**
BEGINRELEASENOTES

* added benchmarking plugin library
* added CMake flag MARLIN_BUILD_BENCHMARKS to build benchmark plugin library
* added script for running multiple configurations of Histogram Benchmarks
* rename run-benchmarks to run-benchmarks-scheduler
* added processors for Marlin v2_01_pre for comparison

* fixed race condition in shared hist entry
* reuse shared access buffer (performance boost)

ENDRELEASENOTES

### CODE REVIEW

**Comment by rete** [comment]
WIP ?

**Comment by JBenda** [comment]
yes. :sweat_smile:

**Comment by rete** [comment]
Can you rename the processor to something more like `MarlinBenchHistProcessor` or similar. The reason is that as soon as you create a new processor plugin, the name is taken in the user namespace and can't be called the same anymore.

File: benchmarking/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#################################
+# cmake file for building benchmarking component
+# @author Remi Ete, Benda Julian, DESY
+#################################
+
+# create library
+aux_source_directory( plugins BenchmarkPlugins_sources )
+add_library( BenchmarkPlugins MODULE ${BenchmarkPlugins_sources} )
```
Comment (rete) [inline benchmarking/CMakeLists.txt]: BenchmarkPlugins -> MarlinBenchmarkPlugins

File: benchmarking/plugins/MarlinBenchHistNativProcessor.cc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+void MarlinBenchHistNativeProcessor::processEvent(EventStore * evt) {
+  if (_shared) {
+    std::list<RHistConcurrentFiller<RH1F, 1024>> fillers{}; 
+    for(auto & manager : _managers) {
+      fillers.emplace_back(manager);
+    }
+    for(marlin::size_t i = 0; i < _nFills; ++i) {
+      for ( auto& filler : fillers) {
+        filler.Fill({_distributionV(_generator)}, 1.);
```
Comment (rete) [inline benchmarking/plugins/MarlinBenchHistNativProcessor.cc]: Thread safety for `_distributionV` and `_generator` ?

File: benchmarking/plugins/MarlinBenchHistProcessor.cc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+void MarlinBenchHistProcessor::processEvent(EventStore * evt) {
+  std::vector<H1FHandle> hists{}; 
+  for(auto itr = _histograms.begin(); itr != _histograms.end(); ++itr) {
+    hists.emplace_back(itr->handle());
+  }
+  
+  for(marlin::size_t i = 0; i < _nFills; ++i) {
+    for ( auto& hist : hists) {
+      hist.fill({_distributionV(_generator)}, 1.);
```
Comment (rete) [inline benchmarking/plugins/MarlinBenchHistProcessor.cc]: Same here. Thread safety ?

File: benchmarking/run-benchmarking-bookstore [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+fi
+
+if [ "${triggerUnpacking}" == "true" ]
+then
+  triggerUnpackingStr="trig"
+else
+  triggerUnpackingStr="notrig"
+fi
+
+output="MarlinMTBenchmarking_Lazy.${lazyUnpackStr}_Trigger.${triggerUnpackingStr}_Sigma.${sigmapercent}.txt"
```
Comment (rete) [inline benchmarking/run-benchmarking-bookstore]: Include other options in the file name, like memory layout and native.

File: benchmarking/run-benchmarking-bookstore [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+# load the file in the cache for next
+echo "Loading input file into cache: ${inputfile}"
+anajob ${inputfile} >> /dev/null
+
+if [[ "$native" == "true" ]] && [[ $memorylayout != "Shared" ]]; then
+	echo "ERROR copied memory layout is native not possible!"
+	exit 1
+fi
+echo "$memorylayout native: $nativeShared REMOVE! TODO"
```
Comment (rete) [inline benchmarking/run-benchmarking-bookstore]: TODO ?

File: benchmarking/run-benchmarking-bookstore [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+				exit 1
+			fi
+			serial=$( cat scaling_temp_$$.log | grep serial | awk '{print $7}' )
+			parallel=$( cat scaling_temp_$$.log |  grep serial | awk '{print $9}' )
+			scaling=$( cat scaling_temp_$$.log | grep serial| awk '{print $11}' )
+			echo "Result: "
+			echo "  serial Time:   ${serial}"
+			echo "  parallel Time: ${parallel}"
+			echo "  scaling:       ${scaling}"
+			echo "${c} ${cr} ${crunchSigma} ${nh} ${serial} ${parallel} ${scaling}" >> ${output}
```
Comment (rete) [inline benchmarking/run-benchmarking-bookstore]: What kind of plots do you to produce out of this benchmark ? The data you want to write on this line may depends on the answer

File: test/processors/TestHistogram.cc [hunk 7]
```diff
@@ -22,26 +23,36 @@ class TestHistogram : public Processor {
   void end() final;
 
 private:
-  H1FEntry _histogram;
+  std::array<H1FEntry,2> _histogram;
```
Comment (rete) [inline test/processors/TestHistogram.cc:26]: Nice example of handling arrays of histograms !

**Review by rete — CHANGES_REQUESTED** [review]
You also renamed the file `benchmarking/run-benchmarking-scheduler`. Did you take care of changing other files that reference this file ?

File: benchmarking/plugins/MarlinBenchHistNativProcessor.cc (hunk 2 above)
Comment (JBenda) [reply benchmarking/plugins/MarlinBenchHistNativProcessor.cc]: distribution should be thread-safe,
and for generator Valgrind has noticed a problem.

File: benchmarking/hist_filling.xml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    <parameter name="LCIOInputFiles">  </parameter>
+    <parameter name="MaxRecordNumber" value="0"/>
+    <parameter name="SkipNEvents" value="0"/>
+    <parameter name="LazyUnpack" value="true"/>
+  </datasource>
+
+  
+  <geometry type="EmptyGeometry" />
+
+  <processor name="UnpackEvent" type="LCIOEventUnpacking"/>
```
Comment (rete) [inline benchmarking/hist_filling.xml]: You create the processor but you don't have it in the execute section. Do you want to add it in the execute section or do you want to remove totally from the config ?

File: benchmarking/plugins/MarlinBenchHistNativProcessor.cc [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            "normal distributed values", 
+            static_cast<int>(_nBins), 
+            -20.f, 20.f));
+      _managers.emplace_back(_histograms.back());
+    }
+  }
+}
+
+void MarlinBenchHistNativeProcessor::processEvent(EventStore * evt) {
+  auto seed = ProcessorApi::getRandomSeed( this, evt ) ;
```
Comment (rete) [inline benchmarking/plugins/MarlinBenchHistNativProcessor.cc]: If you get a random seed, you need to register the processor to get a random seed using the correct function from `ProcessorApi`. Call it in the `init` function.

File: benchmarking/plugins/MarlinBenchHistProcessor.cc [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+          this, "/", std::to_string(i), "dummy histogram", 
+          {"normal distributed values", 
+            book::details::safe_cast<int, std::size_t>(_nBins),
+            -50.f, 50.f}));
+    }
+  }
+}
+
+void MarlinBenchHistProcessor::processEvent(EventStore * evt) {
+  auto seed = ProcessorApi::getRandomSeed( this, evt ) ;
```
Comment (rete) [inline benchmarking/plugins/MarlinBenchHistProcessor.cc]: Same thing here for the random seed

**Review by rete — CHANGES_REQUESTED** [review]

**Pull request merged**